### PR TITLE
Added contract and tests for private sale

### PR DESCRIPTION
**Pull request by marko-perajica**

### CODE REVIEW

File: contracts/exchange/EndemicPrivateSale.sol [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external payable nonReentrant {
+        if (deadline < block.timestamp) {
+            revert PrivateSaleExpired();
+        }
+
+        if (price != msg.value) {
```
Comment (mmarinovic) [inline contracts/exchange/EndemicPrivateSale.sol:83]: Let's do `if(msg.value < price)`

File: contracts/exchange/EndemicPrivateSale.sol [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+            abi.encode(
+                EIP712_DOMAIN_TYPEHASH,
+                keccak256(bytes(DOMAIN_NAME)),
+                keccak256(bytes("1")),
+                block.chainid,
+                address(this),
+                SALT_HASH
+            )
+        );
+    }
```
Comment (mmarinovic) [inline contracts/exchange/EndemicPrivateSale.sol:68]: Do we need extra function here? Can we put this login directly inside of __EndemicPrivateSale___init_unchained if we are not reusing it?

File: test/exchange/EndemicPrivateSale.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+          ZERO_ONE_ETHER,
+          LAST_YEAR_TIMESTAMP,
+          RANDOM_V_VALUE,
+          RANDOM_R_VALUE,
+          RANDOM_S_VALUE
+        )
+      ).to.be.revertedWith(PRIVATE_SALE_EXPIRED);
+    });
+
+    it('should fail with price not correct (too much provided)', async function () {
```
Comment (mmarinovic) [inline test/exchange/EndemicPrivateSale.js:78]: Too much provided is ok :)

File: test/exchange/EndemicPrivateSale.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+          RANDOM_R_VALUE,
+          RANDOM_S_VALUE,
+          {
+            value: ZERO_ONE_ETHER,
+          }
+        )
+      ).to.be.revertedWith(PRICE_NOT_CORRECT);
+    });
+
+    it('should fail with invalid signature', async function () {
```
Comment (mmarinovic) [inline test/exchange/EndemicPrivateSale.js:112]: Let's add tests:
- should fail to buy with valid signature and invalid buyer (I can't buy private sale that was meant for you)

File: contracts/exchange/EndemicPrivateSale.sol (hunk 2 above)
Comment (marko-perajica) [reply contracts/exchange/EndemicPrivateSale.sol:68]: Of course, we don't reuse it I just put it separately for sake of better explanatory what we are doing when the initial method is called

File: contracts/exchange/EndemicPrivateSale.sol [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+    // Maps nftContract -> tokenId -> seller -> buyer -> invalidated.
+    mapping(address => mapping(uint256 => mapping(address => mapping(address => bool))))
+        private privateSaleInvalidated;
+
+    event PrivateSaleSuccess(
+        address indexed nftContract,
+        address indexed seller,
+        address buyer,
+        uint256 indexed tokenId,
```
Comment (mmarinovic) [inline contracts/exchange/EndemicPrivateSale.sol:48]: What about changing order of these? nftContract, tokenId, seller, buyer?

File: contracts/exchange/EndemicPrivateSale.sol [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        _distributeFunds(
+            price,
+            makerCut,
+            totalCut,
+            royaltieFee,
+            royaltiesRecipient,
+            seller
+        );
+
+        privateSaleInvalidated[nftContract][tokenId][seller][buyer] = true;
```
Comment (mmarinovic) [inline contracts/exchange/EndemicPrivateSale.sol:141]: Let's invalidate before doing any calculations. I would move this to LN 122

**Pull request merged**